### PR TITLE
Add BIN, entrypoint field for kubectl 1.12

### PR DESCRIPTION
**Pull request by lbb**
This CL adds a bin field.
Bin specifies the entry point for a kubectl plugin after the 1.12 release.

### CODE REVIEW

File: pkg/environment/environment.go [hunk 1]
```diff
@@ -65,33 +69,24 @@ func MustGetKrewPathsFromEnvs(envs []string) KrewPaths {
...
 		Download: filepath.Join(os.TempDir(), "krew"),
+		Bin:      filepath.Join(base, "bin"),
 	}
 }
 
-func getKubectlPluginsPath(envs []string) string {
-	envvars := parseEnvs(envs)
-
-	// Look for "${KUBECTL_PLUGINS_PATH}"
-	if path, ok := envvars["KUBECTL_PLUGINS_PATH"]; ok && path != "" {
```
Comment (ahmetb) [inline pkg/environment/environment.go]: Does the implementation completely ignore KUBECTL_PLUGINS_PATH now? (I'm just trying to double-check.) The 1.12 release branch docs don't seem to be updated. https://github.com/kubernetes/website/blob/release-1.12/content/en/docs/tasks/extend-kubectl/kubectl-plugins.md

File: pkg/environment/environment.go [hunk 2]
```diff
@@ -65,33 +69,24 @@ func MustGetKrewPathsFromEnvs(envs []string) KrewPaths {
...
-	// Look for "${XDG_DATA_DIRS}/kubectl/plugins"
-	if path := envvars["XDG_DATA_DIRS"]; path != "" {
-		fullPath := filepath.Join(path, "kubectl", "plugins")
-		stat, err := os.Stat(fullPath)
-		if err == nil && stat.IsDir() {
-			return fullPath
-		}
-	}
+func getKubectlPluginsPath(_ []string) string {
+	// envvars := parseEnvs(envs)
```
Comment (ahmetb) [inline pkg/environment/environment.go]: leftover comment

File: pkg/environment/environment.go [hunk 3]
```diff
@@ -65,33 +69,24 @@ func MustGetKrewPathsFromEnvs(envs []string) KrewPaths {
...
+	// // Look for "${KREW_HOMEDIR}"
+	// if path, ok := envvars["KREW_HOMEDIR"]; ok && path != "" {
+	// 	return path
+	// }
 
 	// Golang does no '~' expansion
 	return filepath.Join(homedir.HomeDir(), ".kube", "plugins")
```
Comment (ahmetb) [inline pkg/environment/environment.go]: I think now that plugins don't have to be in ~/.kube/plugins, do you think we should start using ~/.krew ?

People `rm -rf ~/.kube` to reset their kubectl settings all the time. (I did that myself accidentally the other day).

File: pkg/index/validate.go [hunk 4]
```diff
@@ -69,6 +69,9 @@ func (p Platform) Validate() error {
 	if p.Head == "" && p.URI == "" {
 		return fmt.Errorf("head or URI have to be set")
 	}
+	if p.Bin == "" {
+		return fmt.Errorf("bin has to be set")
+	}
```
Comment (ahmetb) [inline pkg/index/validate.go]: hmm what if we also added `if !starswith(p.Bin, "kubectl-")` check?

File: pkg/index/types.go [hunk 5]
```diff
@@ -45,6 +45,11 @@ type Platform struct {
 
 	Selector *metav1.LabelSelector `json:"selector,omitempty"`
 	Files    []FileOperation       `json:"files"`
+
+	// Bin specifies the path to an executable (kubernetes 1.12).
```
Comment (ahmetb) [inline pkg/index/types.go]: > to an executable (kubernetes 1.12).

to

> to the plugin executable.

no need to note 1.12 IMO unless you're also noting krew version.

File: pkg/environment/environment.go (hunk 2 above)
Comment (lbb) [reply pkg/environment/environment.go]: I'd like to keep it for now (dev-phase).
We should decide where to install krew, as it is required to install plugin binaries into the PATH. We should maybe introduce **KREW_PATH** and **KREW_BIN** variable.

**Comment by ahmetb** [comment]
Do you mind rolling up the `docs/*` changes to the same PR? We probably shouldn't forget those.

File: pkg/environment/environment.go [hunk 6]
```diff
@@ -65,33 +69,18 @@ func MustGetKrewPathsFromEnvs(envs []string) KrewPaths {
...
-		fullPath := filepath.Join(path, "kubectl", "plugins")
-		stat, err := os.Stat(fullPath)
-		if err == nil && stat.IsDir() {
-			return fullPath
-		}
-	}
-
+func getKubectlPluginsPath(_ []string) string {
 	// Golang does no '~' expansion
 	return filepath.Join(homedir.HomeDir(), ".kube", "plugins")
```
Comment (ahmetb) [inline pkg/environment/environment.go]: (copying from the disappeared thread)
I think now that plugins don't have to be in ~/.kube/plugins, do you think we should start using ~/.krew ?

People `rm -rf ~/.kube` to reset their kubectl settings all the time. (I did that myself accidentally the other day).

Comment (lbb) [reply pkg/environment/environment.go]: I'd like to keep it for now (dev-phase).
We should decide where to install krew, as it is also required to install plugin binaries into the PATH. We should maybe introduce KREW_PATH and KREW_BIN variable.

Comment (ahmetb) [reply pkg/environment/environment.go]: maybe just convert this to a method that doesn't get any args for now.

Comment (ahmetb) [reply pkg/environment/environment.go]: also maybe if this method is untested, we can just delete it.

**Comment by ahmetb** [comment]
please rebase so the "go test" works now.

**Comment by ahmetb** [comment]
Thanks! What should we check off in #33?

**Comment by lbb** [comment]
The tracking file currently looks good to me 👍

**Pull request merged**